dandi/metadata.py

```diff
@@ -83,3 +88,213 @@ def get_metadata(path):
     meta["nd_types"] = get_neurodata_types(path)
 
     return meta
+
+
+def parse_age(age):
+    """
+    Convert a human-friendly duration string into an ISO 8601 duration
+
+    Parameters
+    ----------
+    age : str
+
+    Returns
+    -------
+    str
+    """
+    m = re.fullmatch(r"(\d+)\s*(y(ear)?|m(onth)?|w(eek)?|d(ay)?)s?", age, flags=re.I)
+    if m:
+        qty = int(m.group(1))
+        unit = m.group(2)[0].upper()
+        return f"P{qty}{unit}"
+    else:
+        raise ValueError(age)
+
+
+def extract_age(metadata):
+    try:
+        dob = ensure_datetime(metadata["date_of_birth"])
+        start = ensure_datetime(metadata["session_start_time"])
+    except (KeyError, TypeError, ValueError):
+        try:
+            duration = parse_age(metadata["age"])
+        except (KeyError, TypeError, ValueError):
+            return ...
+    else:
+        if start < dob:
+            raise ValueError("session_start_time precedes date_of_birth")
+        duration = timedelta2duration(start - dob)
+    return models.PropertyValue(value=duration, unitText="Years from birth")
+
+
+def timedelta2duration(delta):
+    """
+    Convert a datetime.timedelta to ISO 8601 duration format
+
+    Parameters
+    ----------
+    delta : datetime.timedelta
+
+    Returns
+    -------
+    str
+    """
+    s = "P"
+    if delta.days:
+        s += f"{delta.days}D"
+    if delta.seconds or delta.microseconds:
+        sec = delta.seconds
+        if delta.microseconds:
+            # Don't add when microseconds is 0, so that sec will be an int then
+            sec += delta.microseconds / 1000000
+        s += f"T{sec}S"
+    if s == "P":
+        s += "0D"
+    return s
+
+
+def extract_sex(metadata):
+    value = metadata.get("sex", None)
+    if value is not None:
+        value = value.lower()
+        if value in ["m", "male"]:
+            value_id = "http://purl.obolibrary.org/obo/PATO_0000384"
+            value = "Male"
+        elif value in ["f", "female"]:
+            value_id = "http://purl.obolibrary.org/obo/PATO_0000383"
+            value = "Female"
+        elif value in ["unknown"]:
+            value_id = None
+            value = "Unknown"
+        elif value in ["other"]:
+            value_id = None
+            value = "Other"
+        elif value.startswith("http"):
+            value_id = value
+            value = None
+        else:
+            raise ValueError(f"Cannot interpret sex field: {value}")
+        return models.SexType(identifier=value_id, name=value)
+    else:
+        return ...
```

Comment:
@jwodder and @yarikoptic - i added these mappers for sex (above) and species (below).  for the other mappers it would be good to know what's already in the nwb files across datasets.